metadata/metadata.go

```diff
@@ -113,26 +113,56 @@ type mdOutgoingKey struct{}
 
 // NewIncomingContext creates a new context with incoming md attached.
 func NewIncomingContext(ctx context.Context, md MD) context.Context {
-	return context.WithValue(ctx, mdIncomingKey{}, md)
+	return context.WithValue(ctx, mdIncomingKey{}, []MD{md})
 }
 
 // NewOutgoingContext creates a new context with outgoing md attached.
 func NewOutgoingContext(ctx context.Context, md MD) context.Context {
-	return context.WithValue(ctx, mdOutgoingKey{}, md)
+	return context.WithValue(ctx, mdOutgoingKey{}, []MD{md})
+}
+
+// AppendToOutgoingContext returns a new context with outgoing md attached
+func AppendToOutgoingContext(ctx context.Context, md MD) context.Context {
+	mds, _ := ctx.Value(mdOutgoingKey{}).([]MD)
+	mds = append(mds, md)
```

Comment:
This one is very subtle, but appending to another slice that may be shared by others can result in big problems.  Consider:

```go
ctx := metadata.NewOutgoingContext(context.Background(), md)
ctx2  := metadata.AppendToOutgoingContext(ctx, md2)
ctx3  := metadata.AppendToOutgoingContext(ctx, md3)
```

The operation to append to `ctx3` could potentially destroy ctx2's md2 entry.  You can see that behavior in this example: https://play.golang.org/p/-nTBJQBzyt__g

The fix is to not append to the existing slice.

Reply:
Ahhh. Super good to know, thanks. I'll submit a patch to address this

Reply:
@dfawley I've looked into this behavior of `append` a bit more, and although I don't 100% understand it yet, it seems to be that the pointers are carried around and modified (the "not 100% understanding" part is, I don't fully get how far those get propagated and so on). I've believe I've fixed the problem by essentially doing a copy and append; `append([]MD{md}, mds...)`.